reports/spec1.tex

```diff
@@ -2,4 +2,55 @@
 
 \section{M1: Spec 1}
 
-TODO\footnote{JK: explain how we translated things}
+\begin{figure}
+    \centering
+    \includegraphics[width=\textwidth,angle=-90]{ffg-callgraph.pdf}
+    \caption{The callgraph of the 3SF specification}
+    \label{fig:your_label}
+\end{figure}
+
+The first specification is obtained by following the principle of least surprise; to the extent that it is possible, without regard to model checking efficiency, we translate Python code to its \tlap{} counterpart with the closest syntax, preserving semantics.
+
+Despite being a programming language and specification language respectively, Python and \tlap{} overlap somewhat in what they can express. 
+For instance, both allow us to express sets, and then filter those sets by a given predicate to obtain a different set.
+To take a concrete example, observe the definition of $\mathsf{pset\_filter}$ in Figure~\ref{py_filter}.
+\begin{figure}
+\begin{lstlisting}[language=Python]
+def pset_filter(p:Callable[[T1],bool],s: PSet[T1]) -> PSet[T1]:
+    r: PSet[T1] = pset()
+
+    for e in s:
+        if p(e):
+            r = r.add(e)
+
+    return r
+\end{lstlisting}
+\caption{$\mathsf{pset\_filter}$ definition \label{py_filter}}
+\end{figure}
+%
+Notice that, while filter is not one of the built-in operators of the \emph{pyrsistent} library, it is relatively simple to define a filter operator, such that the set returns contains exactly all of the elements of $s$, for which the Boolean predicate $p$ holds true.
+In \tlap{}, however, filtering is a language primitive, so if we can translate a python set $s$ to a \tlap{} set $\hat{s}$, and a Python predicate $p$ (of the above type) to a \tlap{} predicate $\hat{p}$, we can translate $\mathsf{pset\_filter}(p, s)$ to $\{ x \in \hat{s}\colon \hat{p}(x) \}$.
+
+We take this idea, and apply it to every definition in the file \emph{pythonic\_code\_generic.py}, attempting to identify \tlap{}-equivalents (w.r.t. semantics) for each defined function. 
+Later on, in milestone 3, we give a formal characterization of all of these equivalencies, in the form of rewriting rules, although for the purposes of this first specification, the entirety of this translation is manual.
+
+Here is an example of a top-level operator, $\mathsf{has\_block\_hash}$\footnote{Python definition available here: \url{https://github.com/saltiniroberto/ssf/blob/ad3ba2c21bc1cd554a870a6e0e4d87040558e129/high_level/common/helpers.py\#L20-L24}}, as it is expressed in \tlap{}:
+\begin{lstlisting}[language=tla,columns=fullflexible]
+has_block_hash(block_hash, node_state) ==
+    block_hash \in DOMAIN node_state.view_blocks
+\end{lstlisting}
+%
+We can see that the original Python implementation calls $\mathsf{pmap\_has}$, which checks whether a key is contained in a given map. In our translation, the map is modeled with a \tlap{} function, which has an associated $\DOMAIN$ set, so we simply check for set membership with the built-in $\backslash\mathsf{in}$ operator.
```

Comment:
I haven't compiled this. What is `\DOMAIN`?

Reply:
Its just the capitalized `DOMAIN` operator from tla